sw/device/tests/usbdev_test.c

```diff
@@ -119,12 +120,31 @@ bool test_main(void) {
   usb_testutils_controlep_init(&usbdev_control, &usbdev, 0, config_descriptors,
                                sizeof(config_descriptors), test_descriptor,
                                sizeof(test_descriptor));
-  while (usbdev_control.device_state != kUsbTestutilsDeviceConfigured) {
+
+  // Proceed only when the device has been configured and the timeout has
+  // occurred; this allows host-side software to establish communications.
+  uint32_t timeoutUsecs = 0;
+  if (kDeviceType == kDeviceFpgaCw310) {
+    timeoutUsecs = 5 * 1000000;
+  }
+  ibex_timeout_t timeout = ibex_timeout_init(timeoutUsecs);
+  while (usbdev_control.device_state != kUsbTestutilsDeviceConfigured ||
+         !ibex_timeout_check(&timeout)) {
```

Comment:
Just checking, but this "timeout" (more like a sleep, really) became necessary because the test chooses to send data first? In other words, it doesn't have a control for initiating the character send, since our user space test program doesn't apply configuration. The kernel could configure the device, and we'd get dropped characters before our user space program has opened the tty device to listen.

Sleeps can be rather fragile for synchronization purposes, though. 5 seconds _ought_ to be long enough, so I guess we can punt that. In the future, we might consider using the UART (or a special USB control transaction) to signal our test program to start here.

Reply:
Thanks for prodding me to think about and experiment with that a bit more. I've removed the timeout.

I had convinced myself that without waiting the test could fail because a CI machine would not have time to detect and establish the serial connection before our transmission. What actually happens is that the usbdev/simpleserial together just retain our short "Hi!Hi!" message until the 'cat /dev/ttyUSBx' command collects the data.